raft-tech/TANF-app
# Automatic Generation of OpenAPI documentation

### This pull request changes...

Adds automatic generation of OpenAPI spec documentation for all backend endpoints using the [drf-yasg2](https://pypi.org/project/drf-yasg2/) package.

Additionally, to prevent existing pip packages from being updated to the latest versions when adding this new dependency all of the versions in `Pipfile` were pinned to explicitly comparable versions (~=x.x.x). This circumvents a problem introduced when `PyJWT` is upgraded to v2.0.1 which results in authentication failures.

## FUNCTIONALITY
Addresses https://github.com/raft-tech/TANF-app/issues/564

As is this exposes documentation at a few different URL endpoints, which can be further tuned as desired by the Team. We can either leave all of these in place or only expose one/some of them.

## ReDocs
Available at `/redocs/` - Presents the documentation in a three-panel ReDocs format.

![Screenshot from 2021-01-26 17-54-53](https://user-images.githubusercontent.com/22626085/105916897-a84f4a80-5fff-11eb-99ef-23b7c6596e63.png)

## Swagger UI
Available at `/swagger/` - presents the documentation in Swagger UI format

![Screenshot from 2021-01-26 17-56-17](https://user-images.githubusercontent.com/22626085/105917003-d2087180-5fff-11eb-9baa-5a9473e6a840.png)

## Raw OpenAPI spec
Available at `/swagger.json` or `/swagger.yaml` - presents the OpenAPI specification in either JSON or YAML

## TO TEST

* Navigate to `http://localhost:8080/redocs/` and confirm ReDocs format documentation is available and correctly generated
* Navigate to `http://localhost:8080/swagger/` and confirm Swagger UI style documentation is available and correctly generated
* Navigate to `http://localhost:8080/swagger.[json|yaml]` and confirm OpenAPI spec is generated in the appropriate format.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] API documentation must be automatically generated for all exposed endpoints
  - [x] Documentation must meet OpenAPI standards
  - [x] Generated documentation must be convertable to HTML format so that it can be served in a browser

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
